### PR TITLE
Loss scale support for chainerx 

### LINKED ISSUE #7756: Support loss_scale on chainerx backprop

### PR DESCRIPTION
**Pull request by emcastillo**
Reworks #7763
Closes #7756

Instead of exposing a `loss_scale` attribute on Chainerx arrays. 
We do the loss scaling on the arrays that belong to input variables, parameters or must be explicitly retained at the end of the ChainerX backward pass.

### CODE REVIEW

File: chainerx_cc/chainerx/backward.cc [hunk 1]
```diff
@@ -208,13 +208,15 @@ class BackwardImpl {
             const BackpropId& backprop_id,
             DoubleBackpropOption double_backprop,
             std::unordered_map<ArrayNode*, internal::GradRef> array_node_grad_map,
-            bool retain_grad)
+            bool retain_grad,
+            const absl::optional<float>& loss_scale)
```
Comment (asi1024) [inline chainerx_cc/chainerx/backward.cc]: How about using `float loss_scale` defaults to `1.0`?

File: chainer/function_node.py [hunk 2]
```diff
@@ -1014,23 +1014,23 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
...
-
         # Need to access the arrays to invoke the chainer grad function
         if grad_outputs:
             grad_outputs_chx = [x._data[0] for x in grad_outputs]
         else:
             grad_outputs_chx = []
         outputs_chx = [x._data[0] for x in outputs]
         inputs_chx = [x._data[0] for x in inputs]
+        # pybind has issues when converting opt<int> -> opt<float>
+        if loss_scale:
```
Comment (asi1024) [inline chainer/function_node.py]: It seems to be evaluated to `False` when `loss_scale` is an integer `0`.
```suggestion
        if loss_scale is not None:
```

File: tests/chainer_tests/test_function_node.py [hunk 3]
```diff
@@ -1033,6 +1026,34 @@ def expected_double_grad(self):
         return [ggrad]
 
 
+@testing.parameterize(*testing.product({
+    'loss_scale': [None, 1, 10],
```
Comment (asi1024) [inline tests/chainer_tests/test_function_node.py]: How about adding also float value?

File: tests/chainer_tests/test_variable.py [hunk 4]
```diff
@@ -2824,13 +2824,14 @@ def setUp(self):
         self.x = np.random.uniform(-1, 1, self.in_shape).astype(self.dtype)
         self.y = np.random.uniform(-1, 1, self.in_shape).astype(self.dtype)
 
-    def check_loss_scale(self, x_data, y_data):
+    def check_loss_scale(self, x_data, y_data, chainerx=False):
```
Comment (asi1024) [inline tests/chainer_tests/test_variable.py]: It shadows the global variable `chainerx` which have chainerx module, so it seems better to use another variable name.
```suggestion
    def check_loss_scale(self, x_data, y_data, use_chainerx=False):
```
It seems also OK to pass `xp` as an argument.
```suggestion
    def check_loss_scale(self, xp, x_data, y_data):
```

File: chainerx_cc/chainerx/backward.cc (hunk 1 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/backward.cc]: I thought of it, but I would like to disable some actions such as array tracking if the loss scaling is not set, I didn't want to do the checks as `loss_scale != 1.0` as it doesn't seem safe enough.

**Comment by emcastillo** [comment]
@asi1024  I addressed most of the comments but one, can you please take a look,
thanks!

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**